sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/ExternalCatalogUtils.scala

```diff
@@ -63,22 +63,26 @@ object ExternalCatalogUtils {
     bitSet
   }
 
-  def needsEscaping(c: Char): Boolean = {
+  @inline private final def needsEscaping(c: Char): Boolean = {
     c < charToEscape.size() && charToEscape.get(c)
   }
 
   def escapePathName(path: String): String = {
-    val builder = new StringBuilder()
-    path.foreach { c =>
-      if (needsEscaping(c)) {
-        builder.append('%')
-        builder.append(f"${c.asInstanceOf[Int]}%02X")
-      } else {
-        builder.append(c)
+    val firstIndex = path.indexWhere(needsEscaping)
+    if (firstIndex == -1) {
+      path
+    } else {
+      val builder = new StringBuilder(path.substring(0, firstIndex))
+      path.substring(firstIndex).foreach { c =>
+        if (needsEscaping(c)) {
+          builder.append('%')
+          builder.append(f"${c.asInstanceOf[Int]}%02X")
+        } else {
+          builder.append(c)
+        }
       }
+      builder.toString()
     }
-
-    builder.toString()
   }
 
 
```

Comment:
We can probably apply similar perf. optimizations to the `unescapePathName` on the next line below.

Reply:
Will address this in a separate PR if you don't mind